Documentation/botr/type-system.md

```diff
@@ -0,0 +1,97 @@
+# Type System Overview
+
+Author: Michal Strehovsky ([@MichalStrehovsky](https://github.com/MichalStrehovsky)) - 2016
+
+## Introduction
+
+The type system is a major component of CoreRT. It represents the modules, types, methods, and fields within a program and provides higher level services to the type system users that lets them get answers to various interesting questions.
+
+Some of the high level services the type system provides are:
+
+* Loading new types from the metadata
+* Computing set of interfaces implemented by a specific type
+* Computing static and instance field layout (assigning offsets to individual fields)
+* Computing static and instance GC layout of types (identifying GC pointers within object/class data)
+* Computing VTable layout (assigning slots to virtual methods)
+* Deciding whether a type can be stored to a location of another type
+
+Three major themes drive the design of the type system:
+
+1. Low overhead and high performance
+2. Concurrency
+3. Extensibility and reusability
+
+Low overhead is achieved by lazy loading - instead of eagerly populating the types with fields, various attributes, names, etc. these are read on demand from the underlying data source (metadata). Caching is used conservatively.
+
+Where necessary, partial classes, extension methods, and pluggable algorithms are used to achieve goal 3 instead of polymorphism and object hierarchies. The reusability of the type system is at the source level (source-including different sets of files to get different features). This allows extensibility without making sacrifices that would take us away from goal 1.
+
+The type system in its purest form (i.e. without any partial class extensions) tries to avoid introducing concepts that are not defined in the [ECMA-335 specification](http://www.ecma-international.org/publications/standards/Ecma-335.htm). The specification is a suggested pre-requisite reading to this document and provides definitions to various terms used in this document.
+
+## Relationship with metadata
+
+While metadata (such as the file formats described in the ECMA-335 specification) has a close relationship with the type system, there is a clear distinction between these two: the metadata describes how a type looks like on the low level (e.g. what is the base class of the type; or what fields does it have), but the type system builds higher level representations of those (e.g. how many bytes are required to store an instance of the type at runtime).
```

Comment:
Depends on point of view ... for me, metadata is the high level abstract form, and bytes required to store an instance is how the type looks like on the low level.